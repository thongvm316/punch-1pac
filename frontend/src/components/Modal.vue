<template>
  <div
    class="modal"
    :class="{ active: open }"
  >
    <span
      class="modal-overlay"
      @click="toggle"
    />
    <div class="modal-container">
      <div class="modal-header">
        <span
          class="btn btn-clear float-right"
          @click="toggle"
        />
        <h3 class="modal-title">
          {{ titleModal }}
        </h3>
      </div>
      <div class="modal-body">
        <div class="content">
          <slot />
        </div>
      </div>
    </div>
  </div>
</template>

<script>
export default {
  name: 'Modal',

  props: {
    title: {
      type: String,
      required: true
    },

    modalOpen: {
      type: Boolean,
      required: true
    }
  },

  data() {
    return {
      open: this.modalOpen,
      titleModal: this.title
    }
  },

  watch: {
    modalOpen() {
      this.open = this.modalOpen
    },

    title() {
      this.titleModal = this.title
    }
  },

  methods: {
    toggle() {
      this.open = !this.open
      this.$emit('update:modalOpen', this.open)
    }
  }
}
</script>
