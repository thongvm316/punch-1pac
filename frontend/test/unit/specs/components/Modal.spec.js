import { shallowMount } from '@vue/test-utils'
import wrapperOps from '../../supports/wrapper'
import Modal from '@/components/Modal'

const propsData = {
  title: 'Title',
  modalOpen: true
}
const scopedSlots = {
  default: '<p class="default-slot">Modal content</p>'
}
const toggle = jest.spyOn(Modal.methods, 'toggle')

const localWrapperOps = {
  ...wrapperOps,
  propsData,
  scopedSlots
}

describe('Modal.vue', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(Modal, localWrapperOps)
  })

  afterEach(() => { wrapper.vm.$destroy() })

  describe('when Modal was mounted', () => {
    it('should render correctly', () => {
      expect(wrapper.exists()).toBeTruthy()
      expect(wrapper.isVueInstance()).toBeTruthy()
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('when props data was changed', () => {
    it('should change data', async () => {
      const newProps = {
        title: 'New title',
        modalOpen: false
      }
      wrapper.setProps({
        title: newProps.title,
        modalOpen: newProps.modalOpen
      })
      await wrapper.vm.$nextTick()

      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('when toggle method', () => {
    it('should toggle method was called', async () => {
      wrapper.find('.modal .modal-overlay').trigger('click')
      await wrapper.vm.$nextTick()

      expect(toggle).toHaveBeenCalled()
      expect(wrapper.vm.open).toEqual(false)
      expect(wrapper.emitted('update:modalOpen')).toBeTruthy()
      expect(wrapper.emitted('update:modalOpen')[0]).toEqual([false])
    })
  })
})
