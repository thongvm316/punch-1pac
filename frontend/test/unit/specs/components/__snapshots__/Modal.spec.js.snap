// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal.vue when Modal was mounted should render correctly 1`] = `
<div class="modal active"><span class="modal-overlay"></span>
  <div class="modal-container">
    <div class="modal-header"><span class="btn btn-clear float-right"></span>
      <h3 class="modal-title">
        Title
      </h3>
    </div>
    <div class="modal-body">
      <div class="content">
        <p class="default-slot">Modal content</p>
      </div>
    </div>
  </div>
</div>
`;

exports[`Modal.vue when props data was changed should change data 1`] = `
<div class="modal"><span class="modal-overlay"></span>
  <div class="modal-container">
    <div class="modal-header"><span class="btn btn-clear float-right"></span>
      <h3 class="modal-title">
        New title
      </h3>
    </div>
    <div class="modal-body">
      <div class="content">
        <p class="default-slot">Modal content</p>
      </div>
    </div>
  </div>
</div>
`;
